sklearn/linear_model/tests/test_sag.py

```diff
@@ -0,0 +1,710 @@
+    intercept = np.atleast_1d(intercept)
+    weights2 = np.atleast_2d(weights2)
+    intercept2 = np.atleast_1d(intercept2)
+
+    assert_array_almost_equal(weights, clf.coef_, decimal=10)
+    assert_array_almost_equal(intercept, clf.intercept_, decimal=10)
+    assert_array_almost_equal(weights2, clf.coef_, decimal=10)
+    assert_array_almost_equal(intercept2, clf.intercept_, decimal=10)
+
+
+@ignore_warnings
+def test_regressor_matching():
+    n_samples = 10
+    n_features = 5
+
+    np.random.seed(10)
+    X = np.random.random((n_samples, n_features))
+    true_w = np.random.random(n_features)
+    y = X.dot(true_w)
+
+    alpha = 1.
+    n_iter = 100
+    fit_intercept = True
+
+    eta = get_eta(X, alpha, fit_intercept, classification=False)
+    clf = Ridge(fit_intercept=fit_intercept, tol=.00000000001, solver='sag',
+                alpha=alpha * n_samples, max_iter=n_iter)
+    clf.fit(X, y)
+
+    weights1, intercept1 = sag_sparse(X, y, eta, alpha, n_iter=n_iter,
+                                      dloss=squared_dloss,
+                                      fit_intercept=fit_intercept)
+    weights2, intercept2 = sag(X, y, eta, alpha, n_iter=n_iter,
+                               dloss=squared_dloss,
+                               fit_intercept=fit_intercept)
+
+    assert_array_almost_equal(weights1, clf.coef_, decimal=10)
+    assert_array_almost_equal(intercept1, clf.intercept_, decimal=10)
+    assert_array_almost_equal(weights2, clf.coef_, decimal=10)
+    assert_array_almost_equal(intercept2, clf.intercept_, decimal=10)
+
+
+@ignore_warnings
+def test_sag_pobj_matches_logistic_regression():
+    """tests if the sag pobj matches log reg"""
+    n_samples = 100
+    alpha = 1.0
+    max_iter = 20
+    X, y = make_blobs(n_samples=n_samples, centers=2, random_state=0,
+                      cluster_std=0.1)
+
+    clf1 = LogisticRegression(solver='sag', fit_intercept=False, tol=.0000001,
+                              C=1. / alpha / n_samples, max_iter=max_iter,
+                              random_state=10)
+    clf2 = clone(clf1)
+    clf3 = LogisticRegression(fit_intercept=False, tol=.0000001,
+                              C=1. / alpha / n_samples, max_iter=max_iter,
+                              random_state=10)
+
+    clf1.fit(X, y)
+    clf2.fit(sp.csr_matrix(X), y)
+    clf3.fit(X, y)
+
+    pobj1 = get_pobj(clf1.coef_, alpha, X, y, log_loss)
+    pobj2 = get_pobj(clf2.coef_, alpha, X, y, log_loss)
+    pobj3 = get_pobj(clf3.coef_, alpha, X, y, log_loss)
+
+    assert_array_almost_equal(pobj1, pobj2, decimal=4)
+    assert_array_almost_equal(pobj2, pobj3, decimal=4)
+    assert_array_almost_equal(pobj3, pobj1, decimal=4)
+
+
+@ignore_warnings
+def test_sag_pobj_matches_ridge_regression():
+    """tests if the sag pobj matches ridge reg"""
+    n_samples = 100
+    n_features = 10
+    alpha = 1.0
+    n_iter = 50
+    fit_intercept = False
+    np.random.seed(10)
+    X = np.random.random((n_samples, n_features))
```

Comment:
Please do not mutate the singleton RNG but rather use an isolated instance and avoid causing hard to understand side-effects in the test execution.